### ISSUE TITLE
Class type changes after saving and loading

### ISSUE THREAD

**Issue by hgaiser** [tensorflow/tensorflow#37635]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information** 
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Arch Linux
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: 
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.1.0
- Python version: 3.8.1
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
Creating a model containing a certain layer class (in my case `tf.keras.layers.BatchNormalization`) changes after saving and loading the model.

Before loading it is of type 
```python
tensorflow.python.keras.layers.normalization_v2.BatchNormalization
```

Which is the same as `tf.keras.layers.BatchNormalization`, but after loading it is of type:

```python
tensorflow.python.keras.layers.normalization.BatchNormalization
```

Which is not the same.

**Describe the expected behavior**
Expected behavior is that the type of the layer remains unchanged after saving and loading.

**Standalone code to reproduce the issue** 
```python
import tensorflow as tf

model = tf.keras.Sequential([
	tf.keras.layers.BatchNormalization()
])
model.build(input_shape=(1,))
model.save('/tmp/model.h5')

loaded_model = tf.keras.models.load_model('/tmp/model.h5')

# True
print(isinstance(model.layers[0], tf.keras.layers.BatchNormalization))

# False
print(isinstance(loaded_model.layers[0], tf.keras.layers.BatchNormalization))

# AttributeError: module 'tensorflow' has no attribute 'python'
import tensorflow.python.keras.layers.normalization
print(isinstance(loaded_model.layers[0], tensorflow.python.keras.layers.normalization.BatchNormalization))
```

**Other info / logs** Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.
The reason I want to do this is because I want to freeze updating of the batch normalization weights, which can be done by setting `layer.training = False`. My goal is to do something like this:
```python
for l in model.layers:
    if isinstance(l, tf.keras.layers.BatchNormalization):
        l.training = False
```

**Comment by hgaiser** [comment]
I've dug a bit deeper into this issue.

It seems to come from the issue that there are two versions of `BatchNormalization`. There is a `normalization_v2.BatchNormalization` and a `normalization.BatchNormalization`. If tf2 is enabled, v2 should be used; if tf1 is enabled, v1 should be used.

This check should be implemented here: https://github.com/tensorflow/tensorflow/blob/0d7620c3bc50b01054804d8c631b7eed45a7925b/tensorflow/python/keras/layers/serialization.py#L51-L54

If tf2 is enabled, it imports from `normalization_v2`, therefore overriding the global class `BatchNormalization` with the one from `normalization_v2.BatchNormalization`. The problem is that even though `tf2` is enabled, the code path here thinks `tf2` is disabled. The cause is the order in which code is executed. It currently happens like this:

1. `tensorflow_core/__init__.py` is called at some point for the first time, which indirectly imports `tensorflow/python/keras/layers/serialization.py`.
2. [`tensorflow/python/keras/layers/serialization.py`](https://github.com/tensorflow/tensorflow/blob/0d7620c3bc50b01054804d8c631b7eed45a7925b/tensorflow/python/keras/layers/serialization.py#L51-L54) checks to see if tf2 is enabled using [`tensorflow/python/tf2.py`](https://github.com/tensorflow/tensorflow/blob/v1.15.2/tensorflow/python/tf2.py#L43), which [checks `TF2_BEHAVIOR`](https://github.com/tensorflow/tensorflow/blob/v1.15.2/tensorflow/python/tf2.py#L46) environment flag, which defaults to `False`, so it returns False, so the modules are imported as if tf2 is disabled.
3. Later on in [`tensorflow_core/__init__.py`](https://github.com/tensorflow/tensorflow/blob/5c00e793c61860bbf26778cd4704313e867645be/tensorflow/api_template.__init__.py#L98-L100), tf2 functionality is enabled.

The same problem presumably holds for the other layers in the same code block (`recurrent_v2` and `preprocessing.normalization`).

I see two workarounds for now.

1. By far the easiest is to prefix the command with `TF2_BEHAVIOUR=1` so that the environment variable already enables TF2 before anything is imported. This is the one I'll be using for now.
2. Load models using `tf.keras.models.load_model(path, custom_objects={'BatchNormalization': tf.keras.layers.BatchNormalization}`, but this is a bit less convenient.

**Comment by amahendrakar** [comment]
@hgaiser,
I was able to reproduce the issue with [TF2.1](https://colab.research.google.com/gist/amahendrakar/d51763911edd48c97e8c28050d8ebd18/37635.ipynb). However the issue seems to be fixed in [TF-nightly](https://colab.research.google.com/gist/amahendrakar/e5eb5193e41a38359d52fbe0d226240d/37635-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by hgaiser** [comment]
Hey @amahendrakar , thank you for your response. The `AssertionError` isn't the issue, the issue is that after saving and loading a model the type of the layer has changed. Expected output of the code you executed should be:
```shell
True
True
False
```

Note that this also goes wrong with creating models via `tf.keras.applications`:

```python
import tensorflow as tf

model = tf.keras.applications.resnet50.ResNet50()
assert(isinstance(model.layers[3], tf.keras.layers.BatchNormalization)), f"Layer is {model.layers[3]}, but expected {tf.keras.layers.BatchNormalization}."
```

The default BatchNormalization should be the one from tf2. This is broken both in TF2.1 and TF-nightly.

**Comment by goldiegadde** [comment]
@hgaiser  this is similiar to #36700, a fix is underway and should be available in the next rc candidate of TF 2.2.0

**Comment by hgaiser** [comment]
Yes seems identical, I will close this issue then. Thanks !
